// ng-sidebar-accordion/esm2015/lib/sidebar-accordion-content.component.js
import { ChangeDetectionStrategy, Component, HostBinding, Input, Optional } from '@angular/core';
import { SidebarAccordionComponent } from "./sidebar-accordion/sidebar-accordion.component";
export class SidebarAccordionContentComponent {
    constructor(_container) {
        this._container = _container;
        this.classNameSidebarAccordionContent = true;
        if (!this._container) {
            throw new Error('<ng-sidebar-accordion-content> must be inside a <ng-sidebar-accordion></ng-sidebar-accordion>.');
        }
    }
}
SidebarAccordionContentComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-sidebar-accordion-content',
                template: `
    <ng-content></ng-content>`,
                changeDetection: ChangeDetectionStrategy.OnPush
            },] }
];
SidebarAccordionContentComponent.ctorParameters = () => [
    { type: SidebarAccordionComponent, decorators: [{ type: Optional }] }
];
SidebarAccordionContentComponent.propDecorators = {
    classNameSidebarAccordionContent: [{ type: HostBinding, args: ['class.ng-sidebar-accordion-content',] }],
    className: [{ type: Input }, { type: HostBinding, args: ['class',] }]
};
