// ng-sidebar-accordion/esm2015/lib/sidebar-header.component.js
import { ChangeDetectionStrategy, Component, EventEmitter, HostBinding, HostListener, Input, Optional, Output } from '@angular/core';
import { SidebarComponent } from './sidebar.component';
export class SidebarHeaderComponent {
    constructor(_container) {
        this._container = _container;
        this.classNameSidebarHeader = true;
        this.clicked = new EventEmitter();
        this.touchMoved = new EventEmitter();
        this.touchEnded = new EventEmitter();
        if (!this._container) {
            throw new Error('<ng-sidebar-header> must be inside a <ng-sidebar></ng-sidebar>.');
        }
    }
    onHeaderClick(e) {
        this.clicked.emit(e);
    }
    onHeaderTouchMove(e) {
        e.preventDefault();
        this.touchMoved.emit(e);
    }
    onHeaderTouchEnd(e) {
        e.preventDefault();
        this.touchEnded.emit(e);
    }
    ngOnInit() {
        this._container._addHeader(this);
    }
    ngOnDestroy() {
        this._container._removeHeader(this);
    }
}
SidebarHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-sidebar-header',
                template: `
    <div class="ng-sidebar-header__content">
      <div class="ng-sidebar-header__wrapper">
        <ng-content></ng-content>
      </div>
    </div>`,
                changeDetection: ChangeDetectionStrategy.OnPush
            },] }
];
SidebarHeaderComponent.ctorParameters = () => [
    { type: SidebarComponent, decorators: [{ type: Optional }] }
];
SidebarHeaderComponent.propDecorators = {
    classNameSidebarHeader: [{ type: HostBinding, args: ['class.ng-sidebar-header',] }],
    className: [{ type: Input }, { type: HostBinding, args: ['class',] }],
    clicked: [{ type: Output }],
    touchMoved: [{ type: Output }],
    touchEnded: [{ type: Output }],
    onHeaderClick: [{ type: HostListener, args: ['click', ['$event'],] }],
    onHeaderTouchMove: [{ type: HostListener, args: ['touchmove', ['$event'],] }],
    onHeaderTouchEnd: [{ type: HostListener, args: ['touchend', ['$event'],] }]
};
