// ng-sidebar-accordion/esm2015/lib/sidebar-content.component.js
import { ChangeDetectionStrategy, Component, ElementRef, HostBinding, Input, Optional } from '@angular/core';
import { SidebarComponent } from "./sidebar.component";
export class SidebarContentComponent {
    constructor(_container, element) {
        this._container = _container;
        this.element = element;
        this.classNameSidebarContent = true;
        if (!this._container) {
            throw new Error('<ng-sidebar-content> must be inside a <ng-sidebar></ng-sidebar>.');
        }
    }
    ngOnInit() {
        this._container._addContent(this);
    }
    ngOnDestroy() {
        this._container._removeContent(this);
    }
}
SidebarContentComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-sidebar-content',
                template: `
    <ng-content></ng-content>`,
                changeDetection: ChangeDetectionStrategy.OnPush
            },] }
];
SidebarContentComponent.ctorParameters = () => [
    { type: SidebarComponent, decorators: [{ type: Optional }] },
    { type: ElementRef }
];
SidebarContentComponent.propDecorators = {
    classNameSidebarContent: [{ type: HostBinding, args: ['class.ng-sidebar-content',] }],
    className: [{ type: Input }, { type: HostBinding, args: ['class',] }]
};
